### PR TITLE
storage: Replica.{append,applySnapshot} optimizations

### LINKED ISSUE #8548: performance: applySnapshot inefficiently clears previous replica state
From the profile linked in #8544, `Replica.applySnapshot` taking 14% of the cpu, with a decent chunk of timing being spent in `ReplicaDataIterator.Next`. The usage of `ReplicaDataIterator` by `applySnapshot` is as follows:

```
	iter := NewReplicaDataIterator(&desc, batch, false /* !replicatedOnly */)
	defer iter.Close()
	for ; iter.Valid(); iter.Next() {
		if err := batch.Clear(iter.Key()); err != nil {
			return err
		}
	}
```

That is, we loop over all of the existing replica state and call `batch.Clear`. This is the worst case scenario for the Go-based batch implementation. We call `rocksDBBatchBuilder.Clear` only to immediately flush the deletion to the C++ side. We know the iterator is touching distinct keys, so we could operator on a "distinct" batch (i.e. call `batch.Distinct()` prior to this loop).

### LINKED ISSUE #8640: performance: "blind" Replica.append optimization
`Replica.append` is used to append raft log entries both during normal operation and when applying a snapshot. In the later case, we know that no entries exist (because we just cleared all of the replica's data) and could use `MVCCBlindPut`.

Additionally, we could use blind puts in the normal case if we did an initial seek to find the last entry written.

### PR DESCRIPTION
**Pull request by petermattis**
Use a distinct batch for applying much of the snapshot which avoids
unnecessary flushes to the C++ batch.

Use blind puts for writing raft log entries when possible which avoids
doing unnecessary seeks when we know there is no data present.

Fixes #8548.
Fixes #8640.

---

This change is [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/8644)

### CODE REVIEW

**Comment by petermattis** [comment]
@cockroachdb/stability

File: storage/replica_raftstorage.go [hunk 1]
```diff
@@ -638,7 +651,8 @@ func (r *Replica) applySnapshot(
 	// unreplicated keys from the snapshot.
 	unreplicatedPrefix := keys.MakeRangeIDUnreplicatedPrefix(desc.RangeID)
 
-	// Write the snapshot into the range.
+	// Write the snapshot into the range. The KVs and log entries are all written
```
Comment (spencerkimball) [inline storage/replica_raftstorage.go]: Maybe "The KVs and log entries are all written to distinct keys so we can use a distinct batch" should go up where we create the distinct batch above.

**Comment by spencerkimball** [comment]
LGTM

**Comment by bdarnell** [comment]
:lgtm:

Any performance numbers for this change?

---

Review status: 0 of 1 files reviewed at latest revision, 2 unresolved discussions, all commit checks successful.

---

*[storage/replica_raftstorage.go, line 655 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8644#-KPTY50D8N0KVoQP9Xoo:-KPTY50ECY7YC8PFMitB:b-r3ucp8) ([raw file](https://github.com/cockroachdb/cockroach/blob/4935ec189b0415ffd50ddae5d5d521b9a4d9a0ae/storage/replica_raftstorage.go#L655)):*
> ```Go
> 
> 	// Write the snapshot into the range. The KVs and log entries are all written
> 	// to distinct keys so we can use a distinct batch.
> ```

We may Clear a key above and then Put it here. Does that work correctly with a distinct batch? 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/8644#-:-KPTYbYRUK9xB_crgLs0:b65xmsv)*

**Comment by petermattis** [comment]
Review status: 0 of 1 files reviewed at latest revision, 2 unresolved discussions, some commit checks pending.

---

*[storage/replica_raftstorage.go, line 654 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8644#-KPTLonWtZ5EBoo78vRI-r1-654:-KPTYsnFYogtJG-RAt5F:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/4935ec189b0415ffd50ddae5d5d521b9a4d9a0ae/storage/replica_raftstorage.go#L654)):*
<details><summary><i>Previously, spencerkimball (Spencer Kimball) wrote…</i></summary>
> Maybe "The KVs and log entries are all written to distinct keys so we can use a distinct batch" should go up where we create the distinct batch above.
</details>
Done.

---

*[storage/replica_raftstorage.go, line 655 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8644#-KPTY50D8N0KVoQP9Xoo:-KPTZ0o5LGC5RdD81l9_:b2y5ooh) ([raw file](https://github.com/cockroachdb/cockroach/blob/4935ec189b0415ffd50ddae5d5d521b9a4d9a0ae/storage/replica_raftstorage.go#L655)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary>
> We may Clear a key above and then Put it here. Does that work correctly with a distinct batch? 
</details>
Yes, I'll add a comment explaining why. Essentially, we don't do any reads below as the writes are all "blind".

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/8644)*

**Comment by petermattis** [comment]
No concrete performance numbers, though anecdotally, clearing of old keys and writing the raft log entries both take half the time they previously did. 

---

Review status: 0 of 1 files reviewed at latest revision, 2 unresolved discussions, some commit checks pending.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/8644#-:-KPTZvZkB7cYdbvMcISt:bi6rwek)*

**Pull request merged**